### PR TITLE
GH-36375: [Java] Added creating MapWriter in ComplexWriter.

### LINKED ISSUE #36375: Added creating MapWriter in ComplexWriter.
### Describe the enhancement requested

Add new method rootAsMap() to ComplexWriter and implement it in ComplexWriterImpl for supporting Map type. Because currently it's possible to create MapWriter but like a part of list or struct.
Previously in dremio side:

When i trying to return map like @output ComplexWrite
with this code:

```
org.apache.arrow.vector.complex.writer.BaseWriter.MapWriter mapWriter = out.rootAsList().map(false);

mapWriter.startMap();
for (java.util.Map.Entry<java.lang.Integer, java.lang.Integer> element : map.entrySet()) {
mapWriter.startEntry();
mapWriter.key().integer().writeInt((Integer) element.getKey());
mapWriter.value().integer().writeInt((Integer) element.getValue());
mapWriter.endEntry();
}
mapWriter.endMap();
```
It use UnionMapWriter and generate schema like:
`EXPR$0: Map(false)<$data$: Union(Sparse, [1, 39])<struct: Struct<key: Int(32, true) not null, value: Int(32, true) not null> not null, map: Map(false)<entries: Struct<key: Int(32, true) not null, value: Int(32, true)> not null>>>`
But in OutputDerivation impl class where i should create output Complete type

```
List children = Arrays.asList( CompleteType.INT.toField("key", false), CompleteType.INT.toField("value", false));
return new CompleteType(CompleteType.MAP.getType(), CompleteType.struct(children).toField(MapVector.DATA_VECTOR_NAME, false));
```
(This is only one valid case, because MapVector.initializeChildrenFromFields())
return
EXPR$0::map<key::int32, value::int32> I found a place where it start using union - PromotableWriter.promoteToUnion.
And in the end i have

SCHEMA_CHANGE ERROR: Schema changed during projection. Schema was
schema(EXPR$0::map<key::int32, value::int32>)
but then changed to
schema(EXPR$0::map<struct::struct<key::int32, value::int32>, map::map<key::int32, value::int32>>)

### Component(s)

Java

### PR DESCRIPTION
**Pull request by xxlaykxx**
Added new method rootAsMap() to ComplexWriter and implement it in ComplexWriterImpl for supporting map type.
Previously in dremio side:

When i trying to return map like output ComplexWrite
with this code:

org.apache.arrow.vector.complex.writer.BaseWriter.MapWriter mapWriter = out.rootAsList().map(false);

mapWriter.startMap();
for (java.util.Map.Entry<java.lang.Integer, java.lang.Integer> element : map.entrySet()) {
    mapWriter.startEntry();
    mapWriter.key().integer().writeInt((Integer) element.getKey());
    mapWriter.value().integer().writeInt((Integer) element.getValue());
    mapWriter.endEntry();
}
mapWriter.endMap();
It use UnionMapWriter and generate schema like:
EXPR$0: Map(false)<$data$: Union(Sparse, [1, 39])<struct: Struct<key: Int(32, true) not null, value: Int(32, true) not null> not null, map: Map(false)<entries: Struct<key: Int(32, true) not null, value: Int(32, true)> not null>>>
But in OutputDerivation impl class where i should create output Complete type

List<Field> children = Arrays.asList( CompleteType.INT.toField("key", false), CompleteType.INT.toField("value", false));
return new CompleteType(CompleteType.MAP.getType(), CompleteType.struct(children).toField(MapVector.DATA_VECTOR_NAME, false));
(This is only one valid case, because MapVector.initializeChildrenFromFields())
return
EXPR$0::map<key::int32, value::int32> I found a place where it start using union - PromotableWriter.promoteToUnion.
And in the end i have

SCHEMA_CHANGE ERROR: Schema changed during projection. Schema was 
schema(EXPR$0::map<key::int32, value::int32>)
 but then changed to 
schema(EXPR$0::map<struct::struct<key::int32, value::int32>, map::map<key::int32, value::int32>>)
* Closes: #36375

### CODE REVIEW

**Comment by lidavidm** [comment]
CC @davisusanibar

**Comment by kou** [comment]
Could you open a new issue and prepend the issue ID to the PR title like `GR-XXXX: ...`?
See the auto added comment for details: https://github.com/apache/arrow/pull/36351#issuecomment-1611177965

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java [hunk 1]
```diff
@@ -362,125 +355,6 @@ public void listDecimalType() {
...
-            assertEquals(TimeUnit.MICROSECOND, actual.unit);
-            assertEquals(j, actual.value);
-          }
-        }
-      }
-    }
-  }
-
-  @Test
-  public void listFixedSizeBinaryType() throws Exception {
```
Comment (lidavidm) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java:442]: Why are these tests removed?

Comment (xxlaykxx) [reply java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: my mistake  - reverted

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java [hunk 2]
```diff
@@ -1563,4 +1564,108 @@ public void testStructOfList() {
...
+      mapReader.key().next();
+      assertEquals(1, mapReader.key().reader().readInteger().intValue());
+      mapReader.key().next();
+      assertEquals(2, mapReader.key().reader().readInteger().intValue());
+      assertEquals(1, mapReader.value().readInteger().intValue());
+    }
+  }
+
+  @Test
+  public void testMapWithStruckKey() {
```
Comment (lidavidm) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java:1651]: typo: "struck" -> "struct"?

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java
Comment (lidavidm) [inline java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java:1]: These tests don't actually seem to test ComplexWriterImpl?

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java (hunk 2 above)
Comment (xxlaykxx) [reply java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java]: fixed

File: java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java
Comment (xxlaykxx) [reply java/vector/src/test/java/org/apache/arrow/vector/complex/writer/TestComplexWriter.java:1]: Updated tests with using ComplexWriterImpl

**Review by lidavidm — APPROVED** [review]
Thanks!

**Pull request merged**